electrum/storage.py

```diff
@@ -100,6 +98,18 @@ def write(self, data: str) -> None:
         self._file_exists = True
         self.logger.info(f"saved {self.path}")
 
+    def append(self, data: str) -> None:
+        """ append data to encrypted file"""
+        if self.is_encrypted():
+            self.decrypted += data
+            self.write(self.decrypted)
+            return
+        with open(self.path, "r+") as f:
+            f.seek(0, os.SEEK_END)
+            f.write(data)
+            f.flush()
+            os.fsync(f.fileno())
```

Comment:
Can it happen that the OS process crashes in the middle of this, already having some of `data` (but not all) written to disk?
E.g.
the file initially on disk is:
```
{ "foo": [ "bar", "baz" ] }
```
we try to append this diff:
```
[ { "op": "add", "path": "/foo/1", "value": "qux" } ]
```
so we expect to end up with:
```
{ "foo": [ "bar", "baz" ] }
[ { "op": "add", "path": "/foo/1", "value": "qux" } ]
```
but the file will only contain:
```
{ "foo": [ "bar", "baz" ] }
[ { "op": "add", "path": "/fo
```
-----
Basically I have doubts whether we can treat this as an atomic operation.
Consider the effort done atm in `storage.write` to make the writes atomic:
https://github.com/spesmilo/electrum/blob/fbf41b582a151c6bfa74caf0507d3455bddedba9/electrum/storage.py#L82-L101
There (as we rewrite the whole file every time) we first write to a temp file, and the replace the temp and the main files. That approach does not work for appends though.

Reply:
See e.g. https://lwn.net/Articles/789600/ and https://stackoverflow.com/a/61832882
AFAIU writes to disk cannot be assumed to be atomic even with fsync -- metadata changes are atomic on modern filesystems (e.g. ext4) but data changes are not. Though on ZFS, I think, even data changes are atomic. (still, we ofc cannot assume users are using ZFS :P)

Reply:
recovery from incomplete write was added, see `JsonDB.maybe_load_incomplete_data`